File: go/internal/v0/pipeline.go
package v0

import (
	"fmt"
	"sort"
	"strings"
	"time"
)

type GoldenType int8

const (
	GoldenMd GoldenType = iota
	GoldenAction
	GoldenAck
	GoldenFill
)

type GoldenEvent struct {
	TsNs    int64
	Type    GoldenType
	ClOrdID uint64
	Line    string
}

func FormatAction(a StrategyAction) string {
	var b strings.Builder
	b.WriteString(`{"type":"action","ts_ns":`)
	b.WriteString(fmt.Sprintf("%d", a.TsNs))
	b.WriteString(`,"action":"`)
	switch a.Kind {
	case ActionNoop:
		b.WriteString("NOOP")
	case ActionNewOrder:
		b.WriteString("NEW_ORDER")
	case ActionCancel:
		b.WriteString("CANCEL")
	}
	b.WriteByte('"')
	switch a.Kind {
	case ActionNewOrder:
		b.WriteString(fmt.Sprintf(`,"cl_ord_id":%d,"side":"%s","price":%d,"qty":%d`,
			a.ClOrdID, a.Side.String(), a.Price, a.Qty))
	case ActionCancel:
		b.WriteString(fmt.Sprintf(`,"cl_ord_id":%d`, a.ClOrdID))
	}
	b.WriteString(fmt.Sprintf(`,"reason":"%s"}`, a.Reason))
	return b.String()
}

func FormatEngineEvent(e EngineEvent) (GoldenType, uint64, string) {
	switch {
	case e.Ack != nil:
		line := fmt.Sprintf(`{"type":"ack","ts_ns":%d,"cl_ord_id":%d,"status":"%s"`,
			e.Ack.TsNs, e.Ack.ClOrdID, e.Ack.Status.String())
		if e.Ack.Status == AckRejected {
			line += fmt.Sprintf(`,"reason":"%s"`, e.Ack.Reason)
		}
		line += "}"
		return GoldenAck, e.Ack.ClOrdID, line
	case e.Fill != nil:
		line := fmt.Sprintf(`{"type":"fill","ts_ns":%d,"cl_ord_id":%d,"price":%d,"qty":%d,"liquidity":"%s"}`,
			e.Fill.TsNs, e.Fill.ClOrdID, e.Fill.Price, e.Fill.Qty, e.Fill.Liquidity.String())
		return GoldenFill, e.Fill.ClOrdID, line
	default:
		md := e.Md
		line := fmt.Sprintf(`{"type":"md","ts_ns":%d,"instrument":"%s","best_bid":%d,"best_ask":%d,"best_bid_qty":%d,"best_ask_qty":%d}`,
			md.TsNs, md.Instrument, md.BestBid, md.BestAsk, md.BestBidQty, md.BestAskQty)
		return GoldenMd, 0, line
	}
}

func SortGolden(events []GoldenEvent) {
	sort.SliceStable(events, func(i, j int) bool {
		a, b := events[i], events[j]
		if a.TsNs != b.TsNs {
			return a.TsNs < b.TsNs
		}
		if a.Type != b.Type {
			return a.Type < b.Type
		}
		return a.ClOrdID < b.ClOrdID
	})
}

func RunPipeline(ticks []Tick, probe *LatencyProbe) []GoldenEvent {
	engine := NewMatchingEngine()
	gateway := NewOrderGateway(engine)
	strategy := NewStrategy()
	var out []GoldenEvent

	for _, tick := range ticks {
		t0 := time.Now()
		action := strategy.OnTick(tick)

		cl := uint64(0)
		if action.Kind != ActionNoop {
			cl = action.ClOrdID
		}
		out = append(out, GoldenEvent{
			TsNs:    action.TsNs,
			Type:    GoldenAction,
			ClOrdID: cl,
			Line:    FormatAction(action),
		})

		if action.Kind != ActionNoop {
			var events []EngineEvent
			if action.Kind == ActionNewOrder {
				events = gateway.NewOrder(Order{
					ClOrdID: action.ClOrdID,
					Side:    action.Side,
					Price:   action.Price,
					Qty:     action.Qty,
				}, action.TsNs)
			} else {
				events = gateway.Cancel(action.ClOrdID, action.TsNs)
			}
			for _, ev := range events {
				gt, id, line := FormatEngineEvent(ev)
				out = append(out, GoldenEvent{
					TsNs:    action.TsNs,
					Type:    gt,
					ClOrdID: id,
					Line:    line,
				})
			}
		}

		probe.RecordNs(time.Since(t0).Nanoseconds())
	}

	SortGolden(out)
	return out
}

File: java/src/main/java/com/ytta/v0/Pipeline.java
package com.ytta.v0;

import com.ytta.v0.Types.AckStatus;
import com.ytta.v0.Types.ActionKind;
import com.ytta.v0.Types.EngineEvent;
import com.ytta.v0.Types.Order;
import com.ytta.v0.Types.StrategyAction;
import com.ytta.v0.Types.Tick;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public final class Pipeline {
    public enum GoldenType {
        MD(0),
        ACTION(1),
        ACK(2),
        FILL(3);

        final int rank;

        GoldenType(int rank) {
            this.rank = rank;
        }
    }

    public record GoldenEvent(long tsNs, GoldenType type, long clOrdId, String line) {}

    public static String formatAction(StrategyAction a) {
        StringBuilder sb = new StringBuilder();
        sb.append("{\"type\":\"action\",\"ts_ns\":").append(a.tsNs()).append(",\"action\":\"");
        switch (a.kind()) {
            case NOOP -> sb.append("NOOP");
            case NEW_ORDER -> sb.append("NEW_ORDER");
            case CANCEL -> sb.append("CANCEL");
        }
        sb.append('"');
        switch (a.kind()) {
            case NEW_ORDER -> sb.append(",\"cl_ord_id\":")
                    .append(a.clOrdId())
                    .append(",\"side\":\"")
                    .append(a.side().wire())
                    .append("\",\"price\":")
                    .append(a.price())
                    .append(",\"qty\":")
                    .append(a.qty());
            case CANCEL -> sb.append(",\"cl_ord_id\":").append(a.clOrdId());
            case NOOP -> {}
        }
        sb.append(",\"reason\":\"").append(a.reason()).append("\"}");
        return sb.toString();
    }

    public static GoldenEvent formatEngineEvent(EngineEvent e, long tsNs) {
        if (e instanceof EngineEvent.Ack ackEv) {
            var ack = ackEv.value();
            StringBuilder sb = new StringBuilder();
            sb.append("{\"type\":\"ack\",\"ts_ns\":")
                    .append(ack.tsNs())
                    .append(",\"cl_ord_id\":")
                    .append(ack.clOrdId())
                    .append(",\"status\":\"")
                    .append(ack.status().wire())
                    .append('"');
            if (ack.status() == AckStatus.REJECTED) {
                sb.append(",\"reason\":\"").append(ack.reason()).append('"');
            }
            sb.append('}');
            return new GoldenEvent(tsNs, GoldenType.ACK, ack.clOrdId(), sb.toString());
        }
        if (e instanceof EngineEvent.Fill fillEv) {
            var fill = fillEv.value();
            return new GoldenEvent(
                    tsNs,
                    GoldenType.FILL,
                    fill.clOrdId(),
                    "{\"type\":\"fill\",\"ts_ns\":"
                            + fill.tsNs()
                            + ",\"cl_ord_id\":"
                            + fill.clOrdId()
                            + ",\"price\":"
                            + fill.price()
                            + ",\"qty\":"
                            + fill.qty()
                            + ",\"liquidity\":\""
                            + fill.liquidity().wire()
                            + "\"}");
        }
        var md = ((EngineEvent.Md) e).value();
        return new GoldenEvent(
                tsNs,
                GoldenType.MD,
                0,
                "{\"type\":\"md\",\"ts_ns\":"
                        + md.tsNs()
                        + ",\"instrument\":\""
                        + md.instrument()
                        + "\",\"best_bid\":"
                        + md.bestBid()
                        + ",\"best_ask\":"
                        + md.bestAsk()
                        + ",\"best_bid_qty\":"
                        + md.bestBidQty()
                        + ",\"best_ask_qty\":"
                        + md.bestAskQty()
                        + "}");
    }

    public static void sortGolden(List<GoldenEvent> events) {
        events.sort(Comparator
                .comparingLong(GoldenEvent::tsNs)
                .thenComparingInt(e -> e.type().rank)
                .thenComparingLong(GoldenEvent::clOrdId));
    }

    public static List<GoldenEvent> run(List<Tick> ticks, LatencyProbe probe) {
        MatchingEngine engine = new MatchingEngine();
        OrderGateway gateway = new OrderGateway(engine);
        Strategy strategy = new Strategy();
        List<GoldenEvent> out = new ArrayList<>();

        for (Tick tick : ticks) {
            long t0 = System.nanoTime();
            StrategyAction action = strategy.onTick(tick);
            long cl = action.kind() == ActionKind.NOOP ? 0 : action.clOrdId();
            out.add(new GoldenEvent(action.tsNs(), GoldenType.ACTION, cl, formatAction(action)));

            if (action.kind() != ActionKind.NOOP) {
                List<EngineEvent> events = switch (action.kind()) {
                    case NEW_ORDER -> gateway.newOrder(
                            new Order(action.clOrdId(), action.side(), action.price(), action.qty()),
                            action.tsNs());
                    case CANCEL -> gateway.cancel(action.clOrdId(), action.tsNs());
                    case NOOP -> List.of();
                };
                for (EngineEvent ev : events) {
                    out.add(formatEngineEvent(ev, action.tsNs()));
                }
            }

            probe.recordNs(System.nanoTime() - t0);
        }

        sortGolden(out);
        return out;
    }
}
